File: ios/keyman/Keyman/Keyman/InfoViewController/InfoViewController.m
//
//  InfoViewController.m
//  Keyman for iPhone and iPad
//
//  Created by Serkan Kurt on 13/06/13.
//  Copyright (c) 2017 SIL International. All rights reserved.
//

#import "InfoViewController.h"
#import "AppDelegate.h"
#import "Keyman.h"
#import "Reachability.h"

@interface InfoViewController ()
    @property (strong, nonatomic) Reachability *networkReachable;
@end

@implementation InfoViewController
@synthesize webView, networkReachable;

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    if ([self respondsToSelector:@selector(setExtendedLayoutIncludesOpaqueBars:)])
        [self setExtendedLayoutIncludesOpaqueBars:YES];
    
    self.webView.delegate = self;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(checkNetworkStatus:) name:kReachabilityChangedNotification object:nil];
	networkReachable = [Reachability reachabilityWithHostName:@"www.keyman.com"];
	[networkReachable startNotifier];
}

- (void)viewDidUnload {
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    self.webView = nil;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)checkNetworkStatus:(NSNotification *)notification {
	NetworkStatus networkStatus = [networkReachable currentReachabilityStatus];
	switch (networkStatus)
	{
		case NotReachable:
		{
			//NSLog(@"Network is Off");
            [self loadFromLocal];
			break;
		}
		default:
		{
			//NSLog(@"Network is On");
            [self loadFromServer];
			break;
		}
	}
}

- (void)loadFromLocal {
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"info" ofType:@"html" inDirectory:nil];
    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL fileURLWithPath:filePath]]];
}

- (void)loadFromServer {
    NSString *currentKbID = [[[KMManager sharedInstance] currentKeyboardInfo] objectForKey:kKeymanKeyboardIdKey];
    if (currentKbID == nil)
        currentKbID = kKeymanDefaultKeyboardID;
    
    NSUserDefaults *userData = [AppDelegate activeUserDefaults];
    NSArray *userKeyboards = [userData arrayForKey:kKeymanUserKeyboardsListKey];
    NSMutableString *installedKbs = [NSMutableString stringWithString:@""];
    NSString *kbID;
    if (userKeyboards != nil) {
        for (NSDictionary *kbInfo in userKeyboards) {
            kbID = [kbInfo objectForKey:kKeymanKeyboardIdKey];
            if ([installedKbs rangeOfString:kbID].location == NSNotFound) {
                [installedKbs appendString:[NSString stringWithFormat:@"%@,", kbID]];
            }
        }
    }
    
    NSInteger lastIndex = [installedKbs length] - 1;
    if (lastIndex > 0)
        [installedKbs deleteCharactersInRange:NSMakeRange(lastIndex, 1)];
    
    if ([installedKbs isEqualToString:@""])
        installedKbs = [NSMutableString stringWithString:currentKbID];
    
    NSString *urlStr = [NSString stringWithFormat:@"http://keyman.com/iphone-and-ipad/app/?active=%@&installed=%@", currentKbID, installedKbs];
    NSURL *url = [NSURL URLWithString:urlStr];
    [webView loadRequest:[NSURLRequest requestWithURL:url]];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
    //NSLog([NSString stringWithFormat:@"Error : %@", error]);
}

- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType {
    if (navigationType == UIWebViewNavigationTypeLinkClicked)
    {
        [[UIApplication sharedApplication] openURL:[request URL]];
        return NO;
    }

    return YES;
}

- (void)reloadKeymanHelp {
    [self checkNetworkStatus:nil];
}

@end

File: ios/keyman/Keyman/Keyman/InfoViewController/InfoViewController.swift
//
//  InfoViewController.swift
//  Keyman
//
//  Created by Gabriel Wong on 2017-09-05.
//  Copyright © 2017 SIL International. All rights reserved.
//

import KeymanEngine
import UIKit
import Reachability

class InfoViewController: UIViewController, UIWebViewDelegate {
  @IBOutlet var webView: UIWebView!

  private var networkReachable: Reachability?

  convenience init() {
    if UIDevice.current.userInterfaceIdiom == .phone {
      self.init(nibName: "InfoViewController_iPhone", bundle: nil)
    } else {
      self.init(nibName: "InfoViewController_iPad", bundle: nil)
    }
  }

  override func viewDidLoad() {
    super.viewDidLoad()
    extendedLayoutIncludesOpaqueBars = true
    webView?.delegate = self
    NotificationCenter.default.addObserver(self, selector: #selector(self.networkStatusChanged),
        name: NSNotification.Name.reachabilityChanged, object: nil)

    do {
      try networkReachable = Reachability(hostname: "keyman.com")
      try networkReachable?.startNotifier()
    } catch {
      log.error("error starting Reachability notifier: \(error)")
    }
  }

  @objc func networkStatusChanged(_ notification: Notification) {
    reloadKeymanHelp()
  }

  func reloadKeymanHelp() {
    let networkStatus = networkReachable?.connection
    switch networkStatus {
    case Reachability.Connection.none?:
      loadFromLocal()
    default:
      loadFromServer()
    }
  }

  private func loadFromLocal() {
    let offlineHelpBundle = Bundle(path: Bundle.main.path(forResource: "OfflineHelp", ofType: "bundle")!)!

    // Safari won't recognize the contents without the .html ending.
    let filePath = offlineHelpBundle.path(forResource: "index", ofType: "html", inDirectory: nil)
    webView.loadRequest(URLRequest(url: URL.init(fileURLWithPath: filePath!)))
  }

  private func loadFromServer() {
    let appVersion = Version.current.majorMinor
    let url = "https://help.keyman.com/products/iphone-and-ipad/\(appVersion.plainString)/"
    webView.loadRequest(URLRequest(url: URL(string: url)!))
    log.debug("Info page URL: \(url)")
  }
}
